distributed/client.py

```diff
@@ -4638,7 +4635,12 @@ class will be instantiated with any extra keyword arguments.
         distributed.WorkerPlugin
         unregister_worker_plugin
         """
-        if isinstance(plugin, type):
+        if isinstance(plugin, type) or kwargs:
+            warnings.warn(
+                "passing plugin classes or kwargs is deprecated; "
+                "pass a constructed plugin instance instead",
+                category=FutureWarning,
+            )
```

Comment:
Sorry if I missed the discussion somewhere, is there a reason to deprecate this feature? It seems useful to me and many other plugin type packages allow passing a class as the plugin so it may be something users expect. Certainly an improved warning message or fixing the broken behavior is warranted.

Reply:
there's a similar PR here: https://github.com/dask/distributed/pull/5394/files#diff-bbcf2e505bf2f9dd0dc25de4582115ee4ed4a6e80997affc7b22122912cc6591R5470 I think it's best to copy that deprecation.

having `register_worker_plugin` take a class and kwargs separately seems to imply that the class will be constructed somewhere else at another time - rather than immediately

Reply:
`register_scheduler_plugin` is interesting because it's different again, the plugin gets instantiated in the event loop thread: https://github.com/dask/distributed/blob/9a66b719d53303b51034b339a97a9131105e5451/distributed/client.py#L4498-L4500

whereas for the other plugin registration methods the plugin is instantiated in the current thread

Reply:
Good catch. Being consistent with when the plugin instance is created would be nice. My guess is either convention is fine. I have a slight preference for creating the plugin inside the event loop thread so that users can be more explicit in `await`ing the registration process.

Reply:
> That's different as it's for methods on the Scheduler, which are much less user-facing than the Client

Well, we're also raising a `FutureWarning` over there so I assume we expect it to be used by someone?

>  I have a slight preference for creating the plugin inside the event loop thread so that users can be more explicit in awaiting the registration process.

I have a preference to do the opposite since we're talking about a loop on Client side here. I don't see a reason why anyone should interact with the client loop at all. Once on the scheduler/worker this is a different story but for the setup there we have appropriate hooks.

When Thomas and I discussed this earlier I was actually thrown off myself for quite some time because I thought this API would only instantiate the class on the remote but that's not the case.

I don't think having two ways to do the same thing is particularly helping with UX, especially not if there is a looming question of "is it executed on the loop or not?" and "is it instantiated on client side or not?"

The bottom line is that I like this change much more than the exception / argument parsing logic we would have in here if we needed to support both use cases. We'd also need much more explicit docs. It also aligns better with the scheduler